components/camel-tahu/src/test/java/org/apache/camel/component/tahu/SparkplugTCKService.java

```diff
@@ -0,0 +1,226 @@
+        LOG.trace("beforeEach called");
+
+        spTckResultMockEndpoint.expectedMessageCount(1);
+        spTckResultMockEndpoint.message(0).body(String.class).contains("OVERALL: PASS");
+
+        spTckResultMockNotify = new NotifyBuilder(monitorCamelContext)
+                .from("direct:" + SparkplugTCKService.SPARKPLUG_TCK_RESULT_TOPIC)
+                .whenCompleted(1)
+                .create();
+
+        spTckLogMockNotify = new NotifyBuilder(monitorCamelContext)
+                .from("direct:" + SparkplugTCKService.SPARKPLUG_TCK_LOG_TOPIC)
+                .filter(PredicateBuilder.and(
+                        // Filter the expected log messages--anything else causes the TCK test to fail
+                        bodyAs(String.class).not().contains("Creating simulated host application"),
+                        bodyAs(String.class).not().contains("Waiting for the Edge and Device to come online"),
+                        bodyAs(String.class).not().contains("Edge Send Complex Data"),
+                        bodyAs(String.class).not().contains("Host Application is online, so using that")))
+                .whenCompleted(1)
+                .create();
+
+        LOG.trace("beforeEach completed");
+    }
+
+    @Override
+    public void afterEach(ExtensionContext extensionContext) {
+        LOG.trace("afterEach called");
+
+        MockEndpoint.resetMocks(monitorCamelContext);
+
+        LOG.trace("afterEach completed");
+    }
+
+    private CamelContext monitorCamelContext;
+    private ProducerTemplate monitorProducerTemplate;
+
+    MockEndpoint spTckLogMockEndpoint;
+    MockEndpoint spTckResultMockEndpoint;
+
+    NotifyBuilder spTckResultMockNotify;
+    NotifyBuilder spTckLogMockNotify;
+
+    @Override
+    public void beforeAll(ExtensionContext extensionContext) {
+        LOG.trace("beforeAll called");
+
+        try {
+            startMonitorCamelContext(extensionContext);
+            startClient(extensionContext);
+        } finally {
+            LOG.trace("beforeAll completed");
+        }
+    }
+
+    private void startClient(ExtensionContext extensionContext) {
+        try {
+            mqttClient = new MqttClient(
+                    getMqttHostAddress(), "Tahu-Test-" + MqttClient.generateClientId(), new MemoryPersistence());
+
+            MqttConnectOptions mqttConnectOptions = new MqttConnectOptions();
+            if (hiveMQService.getUserName() != null) {
+                mqttConnectOptions.setUserName(hiveMQService.getUserName());
+                mqttConnectOptions.setPassword(hiveMQService.getUserPassword());
+            }
+
+            mqttClient.connect(mqttConnectOptions);
+
+            mqttClient.subscribe(SPARKPLUG_TCK_RESULT_TOPIC, spTckMessageListener);
+            mqttClient.subscribe(SPARKPLUG_TCK_LOG_TOPIC, spTckMessageListener);
+
+        } catch (MqttException e) {
+            fail("Exception caught connecting MQTT test client", e);
+        }
+    }
```

Comment:
Do not use `fail` in start up, `setUp` or `tearDown` methods and also don't use `fail` within any code that is providing the test infrastructure. 

It makes it much harder for us to investigate the actual problem of the test when they happen.

Reply:
Removed